lib/src/lib/tokenizer/codeTokenizer.ts

```diff
@@ -32,56 +32,54 @@ export class CodeTokenizer extends Tokenizer {
   }
 
   /**
-   * Runs the parser on a given string. Returns an async iterator returning
-   * tuples containing the stringified version of the token and the
+   * Runs the parser on a given string. Returns a list of Tokens
+   * containing the stringified version of the token and the
    * corresponding position.
    *
    * @param text The text string to parse
    */
-  public *generateTokens(text: string): IterableIterator<Token> {
+  public generateTokens(text: string): Token[] {
     const tree = this.parser.parse(text, undefined, { bufferSize: Math.max(32 * 1024, text.length * 2) });
-    yield* this.tokenizeNode(tree.rootNode);
+    return this.tokenizeNode(tree.rootNode)[0];
   }
 
-  private *tokenizeNode(node: SyntaxNode): IterableIterator<Token> {
-    const fullSpan = new Region(
+  /**
+   * Tokenizes the given node and its child nodes. It will create a list of Tokens
+   * containing the stringified version of the token and the corresponding position.
+   *
+   * @param node The node (and child nodes) that will be tokenized.
+   *
+   * @returns A tuple `(Tokens, (startRow, startCol))`, where `Tokens`
+   * is the current list of tokens, and `(startRow, startCol)` represents
+   * the starting position of the given tokenized node.
+   */
+  private tokenizeNode(node: SyntaxNode): [Token[], [number,number]]{
+    const location = new Region(
       node.startPosition.row,
       node.startPosition.column,
       node.endPosition.row,
       node.endPosition.column
     );
 
-    const location = Region.firstDiff(fullSpan, this.getChildrenRegions(node));
-    assert(location !== null, "There should be at least one diff'ed region");
+    const allNodes = [];
+    allNodes.push(this.newToken("(", location));
+    allNodes.push(this.newToken(node.type, location));
+    for (const child of node.namedChildren) {
 
-    yield this.newToken("(", location);
+      const [childNodes, [childStartRow, childStartCol]] = this.tokenizeNode(child);
 
-    // "(node.type child1 child2 ...)"
-    yield this.newToken(node.type, location);
+      // If the code is already captured in one of the children, the region of the current node can be shortened.
+      if ((childStartRow < location.endRow) || (childStartRow === location.endRow && childStartCol < location.endCol)) {
+        location.endRow = childStartRow;
+        location.endCol = childStartCol;
+      }
 
-    for (const child of node.namedChildren) {
-      yield* this.tokenizeNode(child);
+      allNodes.push(...childNodes);
```

Comment:
This spread operator is probably applied on large lists of `childNodes`, which could have a large performance impact.

In addition, this also copies the items previously added to that list to another list. Passing the tokens then actually becomes a $$\Theta(n^2)$$ operation as some tokens can be moved multiple times.

I suggest passing the array to add to as an argument instead of returning it. This also has the added benefit of allowing a cleaner return type:

```ts
public generateTokens(text: string): Token[] {
    const tree = this.parser.parse(text, undefined, { bufferSize: Math.max(32 * 1024, text.length * 2) });
    const tokens = [];
	tokenizeNode(tree.rootNode, tokens);
	return tokens;
  }

private tokenizeNode(node: SyntaxNode, tokens: Token[]): [number, number] {
	const location = new Region(node.startPosition.row, node.startPosition.column, node.endPosition.row, node.endPosition.column);
	tokens.push(this.newToken("(", location);
	tokens.push(this.newToken(node.type, location);
	for (const child of node.namedChildren) {
		const [childStartRow, childStartCol] = this.tokenizeNode(child, tokens);
		// If the code is already captured in one of the children, the region of the current node can be shortened.
		if ((childStartRow < location.endRow) || (childStartRow === location.endRow && childStartCol < location.endCol)) {
			location.endRow = childStartRow;
			location.endCol = childStartCol;
		}
	}
	tokens.push(this.newToken(")", location);
	return [location.startRow, location.startCol];
}
```

Reply:
This was indeed something I already tested last week, but (surprisingly) didn't have any impact on the execution time.

Although, it indeed does make the return type more clear.